eventlet/green/ssl.py

```diff
@@ -57,11 +58,40 @@ def __init__(self, sock, keyfile=None, certfile=None,
             # this assignment
             self._timeout = sock.gettimeout()
 
-        # nonblocking socket handshaking on connect got disabled so let's pretend it's disabled
-        # even when it's on
-        super(GreenSSLSocket, self).__init__(
-            sock.fd, keyfile, certfile, server_side, cert_reqs, ssl_version,
-            ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
+        if sys.version_info >= (3, 7):
+            # Monkey-patch the sslsocket so our modified self gets
+            # injected into its _create method.
+            def fake_new(self, cls, *args, **kwargs):
+                return self
+
+            orig_new = _original_sslsocket.__new__
+            try:
+                _original_sslsocket.__new__ = fake_new.__get__(self, GreenSSLSocket)
```

Comment:
Do I understand correctly that this patches/unpatches the original SSLSocket class? If that's the case there's gonna be a race condition - if another thread uses SSLSocket after we patch it but before we unpatch it'll break the application if there's no global monkey patching. Unfortunately I have no competing solution ready right now.

Reply:
Yes, you understand it correctly.
That was also a problem that crossed my mind, but I couldn't think of better way since __init__ is no longer a thing for SSLSocket and the _create class method calls __new__ with it's own class as first argument. Because of this, there is no way of overriding the behavior of _create method.
TL;DR: There is no way of making _create use subclass' instance.

Reply:
if to stay with a fake_new, may be it is to assign it much earlier at the initial patch import,
https://github.com/eventlet/eventlet/blob/master/eventlet/patcher.py#L416

Reply:
Wouldn't that just mean higher probability of breaking other apps trying to use SSLsocket?
The less instructions between assignment and reassignment, the lower chance for the race-condition going wrong.